editoast/src/modelsv2/authn.rs

```diff
@@ -0,0 +1,224 @@
+        Self {
+            pool,
+            _role_set: Default::default(),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum AuthDriverError {
+    #[error(transparent)]
+    DieselError(#[from] diesel::result::Error),
+    #[error(transparent)]
+    PoolError(#[from] editoast_models::db_connection_pool::DatabasePoolError),
+}
+
+// TODO: use load_stream instead of load to avoid useless allocations
+
+impl<B: BuiltinRoleSet + Send + Sync> StorageDriver for PgAuthDriver<B> {
+    type BuiltinRole = B;
+    type Error = AuthDriverError;
+
+    #[tracing::instrument(skip_all, fields(%user), ret, err)]
+    async fn ensure_user(&self, user: &UserInfo) -> Result<i64, Self::Error> {
+        let pool = self.pool.clone();
+        let mut conn = pool.get().await?;
+
+        conn.transaction(|conn| {
+            async {
+                let user_id = authn_user::table
+                    .select(authn_user::id)
+                    .filter(authn_user::identity_id.eq(&user.identity))
+                    .first::<i64>(conn)
+                    .await
+                    .optional()?;
+
+                if user_id.is_none() {
+                    tracing::info!("registering new user in db");
+
+                    let id: i64 = dsl::insert_into(authn_subject::table)
+                        .default_values()
+                        .returning(authn_subject::id)
+                        .get_result(conn)
+                        .await?;
+
+                    dsl::insert_into(authn_user::table)
+                        .values((
+                            authn_user::id.eq(id),
+                            authn_user::identity_id.eq(&user.identity),
+                            authn_user::name.eq(&user.name),
+                        ))
+                        .execute(conn)
+                        .await?;
+
+                    return Ok(id);
+                }
+                tracing::debug!("user already exists in db");
+                Ok(user_id.unwrap())
+            }
+            .scope_boxed()
+        })
+        .await
+    }
+
+    #[tracing::instrument(skip_all, fields(%subject_id, %roles_config), ret, err)]
+    async fn fetch_subject_roles(
+        &self,
+        subject_id: i64,
+        roles_config: &RoleConfig<Self::BuiltinRole>,
+    ) -> Result<HashSet<Self::BuiltinRole>, Self::Error> {
+        let mut conn = self.pool.get().await?;
+
+        let roles = authz_role::table
+            .select(authz_role::role)
+            .left_join(
+                authn_group_membership::table.on(authn_group_membership::user.eq(subject_id)),
+            )
+            .filter(authz_role::subject.eq(subject_id))
+            .or_filter(authz_role::subject.eq(authn_group_membership::group))
+            .load::<String>(&mut conn)
+            .await?
+            .into_iter()
+            .map(|role| {
+                Self::BuiltinRole::from_str(role.as_str())
+                    .ok()
```

Comment:
Why the `.ok()`? How about using `.expect()` directly on the `Result`?

Reply:
Because `FromStr` doesn't require it's `Err` to be `Debug`